### PR TITLE
Enum value parsing: do not parse by whitespace

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Addressing https://github.com/vitessio/vitess/issues/15492, this is an initial simple fix that allows whitespace in `enum` value. A more formal fix to follow, as this fix is incomplete in that it still requires more escaping.

A complete fix is likely to benefit from some `sqlparser` capabilities, though the use case goes beyond parsing `CREATE TABLE` statements and thus we might still take a lesser-than-`yacc` approach.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/15492

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schema/parser.go [hunk 1]
```diff
@@ -122,7 +121,7 @@ func parseEnumOrSetTokens(enumOrSetValues string) (tokens []string) {
 		// input should not contain `enum(...)` column definition, just the comma delimited list
 		return tokens
 	}
-	tokens = textutil.SplitDelimitedList(enumOrSetValues)
+	tokens = strings.Split(enumOrSetValues, ",")
```
Comment (mattlord) [inline go/vt/schema/parser.go:124]: Any reason(s) not to use the cvs encoding packing in the stdlib? https://pkg.go.dev/encoding/csv

Comment (dbussink) [reply go/vt/schema/parser.go:124]: The main issue would be that quoting works differently there with `"` and not `'` like in SQL. Also escaping of quotes is different, so we'd have to write something specific for this or leverage instead `sqlparser` as @shlomi-noach mentioned.

**Review by notfelineit — APPROVED** [review]

File: go/vt/schema/parser.go (hunk 1 above)
Comment (mattlord) [reply go/vt/schema/parser.go:124]: Yeah, I did notice that the csv encoder simply *removes* any commas in the values which surprised me. I do think that we should take this opportunity to address this more fully though as MySQL happily accepts enum string values with commas:
```
mysql> create table t1 (id int not null auto_increment primary key, ev enum('hi there', 'with, comma'));
Query OK, 0 rows affected (0.02 sec)

mysql> show create table t1;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                        |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `id` int NOT NULL AUTO_INCREMENT,
  `ev` enum('hi there','with, comma') DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```

Comment (dbussink) [reply go/vt/schema/parser.go:124]: Right, there's a bunch of things that still fail here. I agree we should fix this properly.

Comment (dbussink) [reply go/vt/schema/parser.go]: @shlomi-noach I pushed up a little specific parser for this in https://github.com/vitessio/vitess/pull/15493/commits/f90e797b0294b31d891209787197d91804ec78e8 that deals also with quotes inside the values as well.

I also added some low level test there with stuff like newlines, I think we should also add those to the end-to-end test too to try all the exotic things possible here.

Comment (dbussink) [reply go/vt/schema/parser.go]: Realized my parser was not enough and also added a proper decoder for string values so we handle things line newlines correctly.

File: go/vt/vttablet/onlineddl/vrepl/columns_test.go [hunk 2]
```diff
@@ -333,35 +333,35 @@ func TestGetExpandedColumnNames(t *testing.T) {
 			"expand enum",
 			Column{
 				Type:       EnumColumnType,
-				EnumValues: "'a', 'b'",
+				EnumValues: "'a','b'",
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/vrepl/columns_test.go:336]: Nothing depends on existence of spaces. Part of the fix here is to actually assume they never exist, as you say.

**Review by shlomi-noach — COMMENTED** [review]
Additional changes in `BufDecodeStringSQL` and `parseEnumOrSetTokens` look good to me.

**Comment by shlomi-noach** [comment]
Proposed followup in https://github.com/vitessio/vitess/pull/17133

**Pull request merged**